### PR TITLE
#66: Improved logic to decide if load from source or index

### PR DESCRIPTION
**Pull request by szarnekow**

### CODE REVIEW

**Comment by szarnekow** [comment]
Looks like I cannot invite someone for a review. @mor-n4 please take a look at this PR.

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/editor/DirtyStateAwareLoadFromSourceHelper.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+			return true;
+		}
+		ResourceSetContext context = ResourceSetContext.get(resourceSet);
+		// check that we are not in the builder scope (which has strict boundaries
+		// wrt project dependencies
+		if (!context.isBuilder()) {
+			ImmutableSet<URI> openEditors = ImmutableSet.copyOf(dirtyStateManager.getDirtyResourceURIs());
+			/* ignore project boundaries */
+			return dependsOnAny(resourceURI, openEditors, getIndex(resourceSet), false);
+		}
```
Comment (mor-n4) [inline plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/editor/DirtyStateAwareLoadFromSourceHelper.java]: here we have two invocations of `#dependsOnAny()` (including the one from the super method), which means we traverse & deserialize the dependency graph twice in the worst case.
Would it make sense to somehow combine the URIs from `openEditors` to the set of URIs checked by the super class and have only a single call to `#dependsOnAny()`?

**Comment by mor-n4** [comment]
Thanks, this looks cool!
Great you found a solution that is pretty concise and understandable, considering the complexity of the issue.

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/editor/DirtyStateAwareLoadFromSourceHelper.java (hunk 1 above)
Comment (szarnekow) [inline plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/editor/DirtyStateAwareLoadFromSourceHelper.java]: Yes, that's something we should consider. I'll look into it.

Comment (szarnekow) [inline plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/editor/DirtyStateAwareLoadFromSourceHelper.java]: We won't cross project boundaries in the first case but explicitly do this in the second case. I'd rather like to keep it as is. What do you think?

Comment (mor-n4) [inline plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/editor/DirtyStateAwareLoadFromSourceHelper.java]: ok, makes sense

**Comment by mor-n4** [comment]
builds are "green"
(there was only one test failure; should be fixed by cd6a27dde393f7742e7420b571448ae7e76957d9)

merging now...

**Pull request merged**